src/components/structures/auth/AuthHeaderContext.tsx

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { isEqual } from "lodash";
+import React, {
+    createContext, Dispatch,
+    Fragment,
+    PropsWithChildren,
+    ReactNode,
+    Reducer,
+    ReducerState,
+    useContext,
+    useEffect,
+    useReducer,
+} from "react";
+
+interface AuthHeaderContextValue {
+    title: ReactNode;
+    icon?: ReactNode;
+    hideServerPicker?: boolean;
+}
+
+enum AuthHeaderContextActionType {
+    ADD,
+    REMOVE
+}
+
+interface AuthHeaderContextAction {
+    type: AuthHeaderContextActionType;
+    value: AuthHeaderContextValue;
+}
+
+type AuthContextReducer = Reducer<AuthHeaderContextValue[], AuthHeaderContextAction>;
+
+interface AuthHeaderContextType {
+    state: ReducerState<AuthContextReducer>;
+    dispatch: Dispatch<AuthHeaderContextAction>;
+}
+
+const AuthHeaderContext = createContext<AuthHeaderContextType>(undefined);
+
+export function AuthHeader(content: AuthHeaderContextValue) {
```

Comment:
really this should be `props` given it is being used as a React component?
Maybe also for sanity we should call it something like AuthHeaderModifier, otherwise someone reading the JSX below will think that <AuthHeader renders a header in-place when in reality it'll be elsewhere in the context